### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by kaikai529**
## Description

remake fc (add batch)
modify nn , cl and pl

### CODE REVIEW

File: cnn/convolution_layer.py [hunk 1]
```diff
@@ -36,51 +36,65 @@ def __select_w(self, k_method, in_channel, out_channel, ksize):
...
             sys.stdout.write("Error: The stride is not right\n")
             sys.exit(1)
         # <padding>  Be careful,size of train data change!
         # (padding size is [channel, height+2*pad-k_height/stride, width+2*pad-k_height/stride])
-        padding_data = self.__padding(self.pad, train_data)
+        self.train_data = self.__padding(self.pad, image_data)
         # <convolution>
-        c_result = self.__convolution(padding_data, self.kernel)
+        c_result = self.__convolution(self.train_data, self.kernel)
+        c_result = c_result / np.max(c_result) * 255
```
Comment (Hattomo) [inline cnn/convolution_layer.py]: What does it do ?

File: cnn/convolution_layer.py [hunk 2]
```diff
@@ -74,6 +75,7 @@ def backpropagation(self, input_error):
                         for a in range(in_channel):
                             z[a] = np.sum(y[a])
                         result[j][:, k, l] = z
+        result = result/np.max(abs(result))
```
Comment (Hattomo) [inline cnn/convolution_layer.py]: What does it do ?

File: cnn/main.py [hunk 3]
```diff
@@ -14,27 +14,18 @@
...
-conv_error = conv.backpropagation(error)
-print(conv_error)
+epoch = 100
+for i in range(epoch):
+    conv_out = conv.forwardpropagation(trainData)
+    pool_out = pool.forwardpropagation(conv_out)
+    error = fullc(pool_out, trainLabel)
+    pool_error = pool.backpropagation(error)
+    conv.backpropagation(pool_error)
+    fullc.test(pool_out, trainLabel)
```
Comment (Hattomo) [inline cnn/main.py]:
```suggestion
    conv_out = conv.forwardpropagation(testData)
    pool_out = pool.forwardpropagation(conv_out)
    fullc.test(pool_out,testLabel)
```

File: dnn/neural_network.py [hunk 4]
```diff
@@ -142,7 +141,8 @@ def test(self, test_data, test_label):
         length = len(test_data)
         for i in range(length):
             self.forwardpropagation(test_data[i], False)
-            if self.__compare(test_label[i], self.z[-1]):
+            if (test_label[i] - self.z[-1] < 0.2).all():
+                # if self.__compare(test_label[i], self.z[-1]):
```
Comment (Hattomo) [inline dnn/neural_network.py]: Why you change this?

File: cnn/fully_connenct_layer.py [hunk 5]
```diff
@@ -10,25 +10,19 @@ class Fully_Connect_Layer(nn.Neural_Network):
...
-        error = super().backpropagation(train_data, train_label, isexternal=True)
-        return error.reshape([channel, height, width])
+    def __call__(self, input_data, input_label):
+        (batch, channel, height, width) = np.shape(input_data)
+        up_error = np.zeros([batch, channel * height * width])
+        for i in range(batch):
+            (train_data, train_label) = (input_data[i].flatten(), input_label[i])
+            super().forwardpropagation(train_data)
+            up_error[i] = super().backpropagation(train_data, train_label, isexternal=True)
+        return up_error.reshape(batch, channel, height, width)
```
Comment (Hattomo) [inline cnn/fully_connenct_layer.py]: Why you use __call__

Comment (kaikai529) [reply cnn/fully_connenct_layer.py]: Merged call and train and flatten

File: cnn/main.py (hunk 3 above)
Comment (kaikai529) [reply cnn/main.py]: Each codes is good because cnn structure is not defined.

File: dnn/neural_network.py (hunk 4 above)
Comment (kaikai529) [reply dnn/neural_network.py]: I want to know if dnn fit label.

File: cnn/convolution_layer.py (hunk 1 above)
Comment (kaikai529) [reply cnn/convolution_layer.py]: 正規化

File: cnn/convolution_layer.py (hunk 2 above)
Comment (kaikai529) [reply cnn/convolution_layer.py]: 正規化

**Review by Hattomo — APPROVED** [review]
🎉

**Pull request merged**